README.md
# remko2mqtt
Control your Remko Air Conditioner via MQTT using an ESP8266

# Content
- [Introduction](#introduction)
- [Using the software](#software)
- [Information about the used communication-protocol](#information)
- [Outlook](#outlook)

<a name="introduction"></a>
# Introduction
With this Arduino-project you can control your Remko air conditioner (mainly the MXW, but maybe the SKW, RVT, ATY, ML, RVD or RXT, too) via MQTT and - for example - NodeRED. As the Remko Smart-Control WiFi-Stick only supports devices since 08/2021 I was not able to use this option for my devices, so I programmed this piece of software. At the moment the following device is tested:
- [x] MXW353 (manufactured in 2016)

## Current state of this project:
- [x] Investigation of the original interface-PCB using the switchboard with the original 5-pin-cable (+5V, GND, XT and RCV) together with IR-controller
- [x] Write code to emulate the necessary commands with 1800 Baud
- [x] Create an adapter-board using ESP8266 to emulate original interface-PCB (mainly a simple level-shift from 3V3 to 5V)
- [x] Enhance code to support more than one device addressed via MQTT-topic using multiple GPIOs
- [x] Read-function to receive and evaluate commands of IR-remote-control
- [ ] Discover the used protocol for XYE-connection to use RS485 with address-setting for supporting multiple devices
- [ ] Reverse engineering of the communication-protocol of an original Remko Smart-Control WiFi-Stick to find more commands
- [ ] Enhance the written ESP8266-program with discovered commands of original WiFi-Stick
- [ ] Test more devices (task for the community ;-) )

At the moment the following commands are supported via direct connection of ESP8266 to display-PCB (without using interface-PCB):
- [x] set device on/off
- [x] set operation mode (Auto, heating, cooling, drying)
- [x] set desired temperature (17°C ... 24°C)
- [x] toggle led

Future "nice-to-have" feature will be:
- [ ] bidirectional communication to readout the current temperature and maybe other information about the air-conditioner
- [ ] Follow-Me-Function for remote-temperature (cmd is implemented, but untested. I fear the IR-remote will not send current temperature via RF when the command is sent by another device)

<a name="software"></a>
# Using the software
To use this software with your own Remko-devices you have to compile the software using Arduino with installed ESP8266-support and PubSubClient-library. In the config.h you have to configure the software (IP-Addresses, MQTT-Server, and so on). You can choose between using WiFi or an ethernet-connection using the W5500-ethernet-shield. Furthermore you can select how many Remko-devices you'd like to use (at the moment up to four) and on which GPIO you want to output the commands. The IP-addresses and MQTT-topics can be set within the config.h, too. Finally, I've implemented an OTA-firmware-update-feature that allows you to update the firmware via the webpage. Just open the IP-address of the device with a browser, click on the button "Firmwareupgrade" and follow the instructions. A new firmware-binary can be created using Arduino using "Ctrl+Alt+S".

Via MQTT the current settings will be published. As the communication with the air-conditioner is unidirectional these values get invalid if you are using the original IR-remote-control.

The ESP8266 can be connected via a simple level-shifter-circuit using a BC337 transistor (or similar):
![image](https://user-images.githubusercontent.com/9845353/201296096-8c356c16-f7c0-4b75-8f3c-acfa146ce493.png)

<a name="information"></a>
# Information about the used communication-protocol
Remko uses predefined commands/bit-pattern for each function of their IR-remote controls and uses these static commands/patterns for their interface-PCB, too. The Sino Wealth SH79F161F 8051 microcontroller on the interface-PCB sends these commands, when it detects a change on SWITCH1/2 or ON/OFF inputs, for example.

So each temperature-value of the remote-control has its individual bit-pattern. I've implemented an Arduino-Test-Sketch to send this bit-pattern via a GPIO. I've already implemented the most-important functions (on/off, temperature-setpoint, operation-mode, etc.):
```C
// TurnOn
{0x0, 0xFF, 0xAE, 0xBB, 0xBA, 0xBA, 0xBA, 0xEB, 0xAE, 0xBB, 0xBB, 0xBB, 0xAE, 0xAA, 0xEE, 0xAE, 0xAE, 0xAA, 0xBB, 0xEE, 0xFE}

// TurnOff
{0x0, 0xFF, 0xAE, 0xBB, 0xBA, 0xBA, 0xBA, 0xEB, 0xBA, 0xBB, 0xEB, 0xEE, 0xAA, 0xAE, 0xEE, 0xAE, 0xAA, 0xEA, 0xEE, 0xEE, 0xFE}
```

With this knowledge we can connect an own ESP8266 to the XT-pin instead of using the original interface-PCB (see highlighted connector CN501):

![image](https://user-images.githubusercontent.com/9845353/200952754-aaace814-bc4a-4f64-8142-94f9f48cabbc.png)

Electrical and timing-parameters are:
- On connector CN501 (GND, 5V, XT, REC, 5V) the interface-PCB transmit commands on pin XT
- Standby-voltage is +5V and individual bits toggle between +5V and 0V (TTL-level)
- A command is send two times right after each other
- Each bit takes 550µs, so we have a communication with 1800 baud
- Communication starts with 8 bits (4.4ms) at low level (GND), followed by 8 bits (4.4ms) at high-level (5V)
- Communication ends on high-level (5V)

![image](https://user-images.githubusercontent.com/9845353/200953658-b8c471ee-0560-470b-b5ee-80029d5ff555.png)

<a name="outlook"></a>
# Outlook
It would be great to use the available RS485-connection on CON403 using the XYE-pins and the address-encoder to connect multiple device. Anyway, as the RS485 IC is connected to the microcontroller only unidirectional, we do not have any advantages except an easier connection when running long cables. For the moment the found solution is sufficient for smaller installations with one to four devices within a smaller radius.


CHANGELOG.md
# Changelog
## v1.0.0
- Initial release without receiving

## v1.2.0
- Added code for receiving IR-remote-control-data

## v1.3.0
- Changed Rx-code from timing-based bit-detection to edge-based as the timing of IR-control is bad

config.h
// general software config
const char* versionstring = "v1.3.0 built on ";
const char compile_date[] = __DATE__ " " __TIME__;
const char* host = "remko2mqtt";

#define UseWiFi 1
#define UseStaticIP 1
#define MirrorRxD2TxD 1 // mirrors the received commands to the defined output-pin

// select how many Remko-devices you want to control and on which GPIO
#define RemkoTxDevices 2 // define the numbers of devices you want to control (1...4) - only 3 Devices possible if using W5500 ethernet
#define RemkoRxDevices 2 // define the numbers of devices you want to read (0...4) - only 3 Devices possible if using W5500 ethernet
uint8_t RemkoTxPin[] = {5, 0, 3, 14}; // Enter the desired GPIO pins in this array
uint8_t RemkoRxPin[] = {4, 2, 1, 12}; // Enter the desired GPIO pins in this array
#define RemkoInvertTxD 1 // invert Tx-bits (e.g. when using a transistor with pullup-resistor)
//#define RemkoInvertRxD 1 // invert Rx-bits

#ifdef UseWiFi
  // we are using WiFi connection
  const char* ssid = "MyVeryPrivateWiFiNetwork";
  const char* password = "MyVeryPrivateWiFiNetworkPassword";
#else
  // we are using ethernet-port via W5500

  // Enter a MAC address for your controller below.
  // Newer Ethernet shields have a MAC address printed on a sticker on the shield
  byte mac[] = {0xDE, 0xAD, 0xBE, 0xEF, 0xFE, 0xED}; //MAC
#endif

#ifdef UseStaticIP
  IPAddress local_IP(192, 168, 0, 42);
  IPAddress gateway(192, 168, 0, 1);
  IPAddress subnet(255, 255, 255, 0);
  IPAddress primaryDNS(192, 168, 0, 1); //optional
  IPAddress secondaryDNS(1, 1, 1, 1); //optional
#endif

// setup MQTT
#define mqtt_id "remko2mqtt"
#define mqtt_server "192.168.0.41"
#define mqtt_serverport 1883
#define mqtt_publish_interval 60

#define mqtt_topic_set_debuglevel      "remko/set/debuglevel"
#define mqtt_topic_debug               "remko/debug"
#define mqtt_topic_restart             "remko/restart"

#define mqtt_topic_dev0_set_poweroff   "remko/0/set/poweroff"
#define mqtt_topic_dev0_set_opmode     "remko/0/set/opmode"
#define mqtt_topic_dev0_set_setpoint   "remko/0/set/setpoint"
#define mqtt_topic_dev1_set_poweroff   "remko/1/set/poweroff"
#define mqtt_topic_dev1_set_opmode     "remko/1/set/opmode"
#define mqtt_topic_dev1_set_setpoint   "remko/1/set/setpoint"
#define mqtt_topic_dev2_set_poweroff   "remko/2/set/poweroff"
#define mqtt_topic_dev2_set_opmode     "remko/2/set/opmode"
#define mqtt_topic_dev2_set_setpoint   "remko/2/set/setpoint"
#define mqtt_topic_dev3_set_poweroff   "remko/3/set/poweroff"
#define mqtt_topic_dev3_set_opmode     "remko/3/set/opmode"
#define mqtt_topic_dev3_set_setpoint   "remko/3/set/setpoint"

#define mqtt_topic_dev0_powerstate      "remko/0/powerstate"
#define mqtt_topic_dev0_opmode          "remko/0/opmode"
#define mqtt_topic_dev0_setpoint        "remko/0/setpoint"
#define mqtt_topic_dev1_powerstate      "remko/1/powerstate"
#define mqtt_topic_dev1_opmode          "remko/1/opmode"
#define mqtt_topic_dev1_setpoint        "remko/1/setpoint"
#define mqtt_topic_dev2_powerstate      "remko/2/powerstate"
#define mqtt_topic_dev2_opmode          "remko/2/opmode"
#define mqtt_topic_dev2_setpoint        "remko/2/setpoint"
#define mqtt_topic_dev3_powerstate      "remko/3/powerstate"
#define mqtt_topic_dev3_opmode          "remko/3/opmode"
#define mqtt_topic_dev3_setpoint        "remko/3/setpoint"


Trash/counted_rxd.c
#if RemkoRxDevices >= 1
void ICACHE_RAM_ATTR remko_rxd_isr0() {
  // we detected a start of communication (falling edge)

  // disable interrupt
  detachInterrupt(digitalPinToInterrupt(RemkoRxPin[0]));
    
  // preload with half-bittime (=275us) to read bit in the middle
  lastBitReadTime[0]=micros()-(BitTime/2);

  // start statemachine for reading
  bitcounter_rxd[0]=0;
  remko_readcmd_state[0] = 1; // start reading command
}
#endif

#if RemkoRxDevices >= 2
void ICACHE_RAM_ATTR remko_rxd_isr1() {
  // we detected a start of communication (falling edge)

  // disable interrupt
  detachInterrupt(digitalPinToInterrupt(RemkoRxPin[1]));
    
  // preload with half-bittime (=275us) to read bit in the middle
  lastBitReadTime[1]=micros()-(BitTime/2);

  // start statemachine for reading
  bitcounter_rxd[1]=0;
  remko_readcmd_state[1] = 1; // start reading command
}
#endif

#if RemkoRxDevices >= 3
void ICACHE_RAM_ATTR remko_rxd_isr2() {
  // we detected a start of communication (falling edge)

  // disable interrupt
  detachInterrupt(digitalPinToInterrupt(RemkoRxPin[2]));
    
  // preload with half-bittime (=275us) to read bit in the middle
  lastBitReadTime[2]=micros()-(BitTime/2);

  // start statemachine for reading
  bitcounter_rxd[2]=0;
  remko_readcmd_state[2] = 1; // start reading command
}
#endif

#if RemkoRxDevices >= 4
void ICACHE_RAM_ATTR remko_rxd_isr3() {
  // we detected a start of communication (falling edge)

  // disable interrupt
  detachInterrupt(digitalPinToInterrupt(RemkoRxPin[3]));
    
  // preload with half-bittime (=275us) to read bit in the middle
  lastBitReadTime[3]=micros()-(BitTime/2);

  // start statemachine for reading
  bitcounter_rxd[3]=0;
  remko_readcmd_state[3] = 1; // start reading command
}
#endif

// create array to isr-routine to use for-loops to attach/detach isr
#if RemkoRxDevices == 4
  void (*isr_table[4])(void) = { remko_rxd_isr0, remko_rxd_isr1, remko_rxd_isr2, remko_rxd_isr3};
#elif RemkoRxDevices == 3
  void (*isr_table[4])(void) = { remko_rxd_isr0, remko_rxd_isr1, remko_rxd_isr2, 0};
#elif RemkoRxDevices == 2
  void (*isr_table[4])(void) = { remko_rxd_isr0, remko_rxd_isr1, 0, 0};
#elif RemkoRxDevices == 1
  void (*isr_table[4])(void) = { remko_rxd_isr0, 0, 0, 0};
#else
  void (*isr_table[4])(void) = { 0, 0, 0, 0};
#endif

void remko_rxd_init() {
  for (int device=0;device<RemkoRxDevices;device++){
    remko_rxd_reset(device);
  }
}

void remko_rxd_reset(uint8_t device) {
  // reattach interrupt for next command
  bitcounter_rxd[device]=0;
  remko_readcmd_state[device]=0; // reset statemachine for next command
  #ifdef RemkoInvertRxD
    attachInterrupt(digitalPinToInterrupt(RemkoRxPin[device]), (*isr_table[device]), RISING);
  #else
    attachInterrupt(digitalPinToInterrupt(RemkoRxPin[device]), (*isr_table[device]), FALLING);
  #endif
}

void remko_rxd_step() {
  // read individual bits and evaluate received commands
  for (int device=0; device<RemkoRxDevices; device++) {
    if (micros() - lastBitReadTime[device] >= BitTime) {
      if (remko_readcmd_state[device] == 1) {
        // statemachine is running
        lastBitReadTime[device]+=BitTime; // jump one full bit to the right for next bit
        remko_rxd_readbit(device); // read current bit
      }else if (remko_readcmd_state[device] > 1) {
        // wait 21 bytes (= ignoring second command)
        bitcounter_rxd[device]++;
  
        if (bitcounter_rxd[device]>cmdlength*8) {
          // end of second commmand
          remko_rxd_reset(device);
        }
      }
    }
    
    if (remko_readcmd_state[device] == 2) {
      // we received a new command
      remko_readcmd_state[device] = 3; // acknowledge new command, stop statemachine

      // compare cmd_rxd[device] and remko_cmd to find the received command-type
      int8_t received_cmd_index = -1;
      for (int i=0; i<17; i++) {
        if (memcmp(cmd_rxd[device], remko_cmd, cmdlength) == 0) {
          // we found a matching command
          received_cmd_index = i;
          break;
        }
      }

      switch (received_cmd_index) {
        case 0: // TurnOn
          remko_powerstate[device]=1;
          break;
        case 1: // TurnOff
          remko_powerstate[device]=0;
          break;
        case 2: // Auto
          remko_opmode[device]=0;
          break;
        case 3: // Cool
          remko_opmode[device]=1;
          break;
        case 4: // Dry
          remko_opmode[device]=2;
          break;
        case 5: // Heat
          remko_opmode[device]=3;
          break;
        case 6: // 17
          remko_setpoint[device]=17;
          break;
        case 7: // 18
          remko_setpoint[device]=18;
          break;
        case 8: // 19
          remko_setpoint[device]=19;
          break;
        case 9: // 20
          remko_setpoint[device]=20;
          break;
        case 10: // 21
          remko_setpoint[device]=21;
          break;
        case 11: // 22
          remko_setpoint[device]=22;
          break;
        case 12: // 23
          remko_setpoint[device]=23;
          break;
        case 13: // 24
          remko_setpoint[device]=24;
          break;
        case 14: // FollowMe on
          remko_followme[device]=1;
          break;
        case 15: // FollowMe off
          remko_followme[device]=0;
          break;
        case 16: // Toggle LED
          break;
        default: // unknown command
          break;
      }

      // send the new command to output, if desired
      #ifdef MirrorRxD2TxD
        // copy the received command to the output-buffer
        // this will work with unknown commands, too
        memcpy(cmd_txd[device], cmd_rxd[device], cmdlength);
        // start bitstream-transmission
        bitcounter_txd[device] = 0;
        remko_sendcmd_state[device] = 1;
      #endif
    }
  }
}

void remko_rxd_readbit(uint8_t device) {
  // read current bit from GPIO

  if (bitcounter_rxd[device]<cmdlength*8) {
    // read current bit and write it to cmd
    #ifdef RemkoInvertRxD
      bitWrite(cmd_rxd[device][bitcounter_rxd[device]/8], (bitcounter_rxd[device]-(bitcounter_rxd[device]/8)*8), !digitalRead(RemkoRxPin[device]));
    #else
      bitWrite(cmd_rxd[device][bitcounter_rxd[device]/8], (bitcounter_rxd[device]-(bitcounter_rxd[device]/8)*8), digitalRead(RemkoRxPin[device]));
    #endif
    // check for first two bytes. Should be 0x00 and 0xFF
    if (((bitcounter_rxd[device]==7) && (cmd_rxd[device][0]!=0)) || ((bitcounter_rxd[device]==15) && (cmd_rxd[device][1]!=255))) {
      // unexpected first or second byte
      remko_rxd_reset(device);
      return;
    }

    // increase bitcounter_rxd[device] for next bit    
    bitcounter_rxd[device]++;
  }else{
    // we reached end of the current command
    remko_readcmd_state[device]=2;
    bitcounter_rxd[device]=0;
  }
}


remkocmd.h
const byte cmd_turnon[] =      {0x0, 0xFF, 0xAE, 0xBB, 0xBA, 0xBA, 0xBA, 0xEB, 0xAE, 0xBB, 0xBB, 0xBB, 0xAE, 0xAA, 0xEE, 0xAE, 0xAE, 0xAA, 0xBB, 0xEE, 0xFE};
const byte cmd_turnoff[] =     {0x0, 0xFF, 0xAE, 0xBB, 0xBA, 0xBA, 0xBA, 0xEB, 0xBA, 0xBB, 0xEB, 0xEE, 0xAA, 0xAE, 0xEE, 0xAE, 0xAA, 0xEA, 0xEE, 0xEE, 0xFE};
const byte cmd_auto[] =        {0x0, 0xFF, 0xAE, 0xBB, 0xBA, 0xBA, 0xBA, 0xEB, 0xAA, 0xBB, 0xBB, 0xBB, 0xBB, 0xAA, 0xEA, 0xBA, 0xEA, 0xAE, 0xEB, 0xEE, 0xFE};
const byte cmd_cool[] =        {0x0, 0xFF, 0xAE, 0xBB, 0xBA, 0xBA, 0xBA, 0xEB, 0xAE, 0xBB, 0xBB, 0xBB, 0xAE, 0xAA, 0xEA, 0xAA, 0xBA, 0xEB, 0xEE, 0xEE, 0xFE};
const byte cmd_dry[] =         {0x0, 0xFF, 0xAE, 0xBB, 0xBA, 0xBA, 0xBA, 0xEB, 0xAA, 0xBB, 0xBB, 0xBB, 0xBB, 0xAA, 0xEA, 0xEA, 0xEA, 0xAE, 0xBB, 0xEE, 0xFE};
const byte cmd_heat[] =        {0x0, 0xFF, 0xAE, 0xBB, 0xBA, 0xBA, 0xBA, 0xEB, 0xAE, 0xBB, 0xBB, 0xBB, 0xAE, 0xAA, 0xEA, 0xBA, 0xAB, 0xBB, 0xAE, 0xEE, 0xFE};
const byte cmd_17[] =          {0x0, 0xFF, 0xAE, 0xBB, 0xBA, 0xBA, 0xBA, 0xEB, 0xAE, 0xBB, 0xBB, 0xBB, 0xAE, 0xAA, 0xAA, 0xEE, 0xEA, 0xEE, 0xAE, 0xEE, 0xFE};
const byte cmd_18[] =          {0x0, 0xFF, 0xAE, 0xBB, 0xBA, 0xBA, 0xBA, 0xEB, 0xAE, 0xBB, 0xBB, 0xBB, 0xAE, 0xAA, 0xAA, 0xBB, 0xAB, 0xBB, 0xAB, 0xEE, 0xFE};
const byte cmd_19[] =          {0x0, 0xFF, 0xAE, 0xBB, 0xBA, 0xBA, 0xBA, 0xEB, 0xAE, 0xBB, 0xBB, 0xBB, 0xAE, 0xAA, 0xEA, 0xEE, 0xAE, 0xEE, 0xAA, 0xEE, 0xFE};
const byte cmd_20[] =          {0x0, 0xFF, 0xAE, 0xBB, 0xBA, 0xBA, 0xBA, 0xEB, 0xAE, 0xBB, 0xBB, 0xBB, 0xAE, 0xAA, 0xEA, 0xBA, 0xAB, 0xBB, 0xAE, 0xEE, 0xFE};
const byte cmd_21[] =          {0x0, 0xFF, 0xAE, 0xBB, 0xBA, 0xBA, 0xBA, 0xEB, 0xAE, 0xBB, 0xBB, 0xBB, 0xAE, 0xAA, 0xBA, 0xEB, 0xAE, 0xAE, 0xAE, 0xEE, 0xFE};
const byte cmd_22[] =          {0x0, 0xFF, 0xAE, 0xBB, 0xBA, 0xBA, 0xBA, 0xEB, 0xAE, 0xBB, 0xBB, 0xBB, 0xAE, 0xAA, 0xBA, 0xBB, 0xBB, 0xBA, 0xAA, 0xEE, 0xFE};
const byte cmd_23[] =          {0x0, 0xFF, 0xAE, 0xBB, 0xBA, 0xBA, 0xBA, 0xEB, 0xAE, 0xBB, 0xBB, 0xBB, 0xAE, 0xAA, 0xBA, 0xEE, 0xAE, 0xAE, 0xAB, 0xEE, 0xFE};
const byte cmd_24[] =          {0x0, 0xFF, 0xAE, 0xBB, 0xBA, 0xBA, 0xBA, 0xEB, 0xAE, 0xBB, 0xBB, 0xBB, 0xAE, 0xAA, 0xBA, 0xBA, 0xAB, 0xEB, 0xAE, 0xEE, 0xFE};
const byte cmd_followmeon[] =  {0x0, 0xFF, 0xAE, 0xBB, 0xEB, 0xEA, 0xAA, 0xEB, 0xBA, 0xAE, 0xEA, 0xBA, 0xEE, 0xEE, 0xBA, 0xBA, 0xBB, 0xAE, 0xBB, 0xEA, 0xFE};
const byte cmd_followmeoff[] = {0x0, 0xFF, 0xAE, 0xBB, 0xBA, 0xBA, 0xBA, 0xEB, 0xAE, 0xBB, 0xBB, 0xBB, 0xAE, 0xAA, 0xBA, 0xBA, 0xAB, 0xEB, 0xAE, 0xEE, 0xFE};
const byte cmd_led[] =         {0x0, 0xFF, 0xAE, 0xBB, 0xAE, 0xEB, 0xEA, 0xBA, 0xEE, 0xEE, 0xBA, 0xAE, 0xEA, 0xBA, 0xAE, 0xAB, 0xEB, 0xBA, 0xEE, 0xBA, 0xFE};
